OpenRA.Mods.Common/Traits/Buildings/RepairableBuilding.cs

```diff
@@ -150,11 +153,50 @@ void ITick.Tick(Actor self)
 				// The cost is the same regardless of the amount of people repairing
 				var hpToRepair = Math.Min(Info.RepairStep, health.MaxHP - health.HP);
 
-				// Cast to long to avoid overflow when multiplying by the health
-				var cost = Math.Max(1, (int)(((long)hpToRepair * Info.RepairPercent * buildingValue) / (health.MaxHP * 100L)));
+				// Have all cost values multiplied by 100, cost tends to be miniscule. 100x comes from Info.RepairPercent
+				// Cast to long to avoid overflow
+				var cost = (int)((long)hpToRepair * Info.RepairPercent * buildingValue / health.MaxHP);
 
-				// TakeCash will return false if the player can't pay, and will stop him from contributing this Tick
-				var activePlayers = Repairers.Count(player => player.PlayerActor.Trait<PlayerResources>().TakeCash(cost, true));
+				// We are using the highest possible repair cost to find the players that
+				// can pay it as we cannot know what the actual cost will be
+				var maxMultiplier = 0;
+				for (var cnt = 0; cnt < Repairers.Count; cnt++)
+				{
+					var multiplier = Info.RepairBonuses[cnt] / (cnt + 1);
+					if (multiplier > maxMultiplier)
+						maxMultiplier = multiplier;
+				}
+
+				activePlayers = Repairers.Count(player => player.PlayerActor.Trait<PlayerResources>().CanTakeCash(Math.Max(1, (int)(builtUpCost + cost * maxMultiplier / 100))));
+
+				// Reduce or increase cost based on repair bunuses and amount people repairing
+				cost /= Info.RepairBonuses[activePlayers - 1] / 100;
```

Comment:
Is this even correct, above we multiply and don't divide by the repair bonus.

Reply:
Also, this will crash if there are 0 active players.

Reply:
> Is this even correct, above we multiply and don't divide by the repair bonus.

It is correct. As you can see I use `/=` here and not `=`

> Also, this will crash if there are 0 active players.

moved the check upwards

Reply:
Above we calculate `cost * maxMultiplier / 100`, here we do `cost / multiplier / 100`?

Reply:
above we calculate cost * multiplier / player count
and below I've made the mistake of / by multiplier. I had fixed it to be * multiplier / player count as well